packages/extension-core/src/domains/balances/pool.ts

```diff
@@ -0,0 +1,875 @@
+    } catch (e) {
+      log.error("Failed to persist balances in pool", e)
+    }
+  }
+
+  get balances() {
+    return this.#pool.getValue()
+  }
+
+  get observable() {
+    return this.#pool.asObservable()
+  }
+
+  private updatePool(updates: BalanceJson[]) {
+    const updatesWithIds = new Balances(updates)
+    const existing = this.balances
+    // update initialising set here - before filtering out zero balances
+    // while this may include stale balances, the important thing is that the balance is no longer "initialising"
+    updates.forEach((b) => this.#initialising.delete(`${b.tokenId}:${b.address}`))
+
+    const newlyZeroBalances: string[] = []
+    const changedBalances = Object.fromEntries(
+      updatesWithIds.each
+        .filter((newB) => {
+          const isZero = newB.total.tokens === "0"
+          // Keep new balances which are not zeros
+          const existingB = existing[newB.id]
+          if (!existingB && !isZero) return true
+
+          const hasChanged = !isEqual(existingB, newB.toJSON())
+          // Collect balances now confirmed to be zero separately, so they can be filtered out from the main set
+          if (existingB && hasChanged && isZero) newlyZeroBalances.push(newB.id)
+
+          // Keep changed balances, which are not known zeros
+          return hasChanged && !isZero
+        })
+        .map((b) => [b.id, b.toJSON()])
+    )
+
+    if (Object.keys(changedBalances).length === 0 && newlyZeroBalances.length === 0) return
+
+    const nonZeroBalances =
+      newlyZeroBalances.length > 0
+        ? Object.fromEntries(
+            Object.entries(existing).filter(([id]) => !newlyZeroBalances.includes(id))
+          )
+        : existing
+    const newBalancesState = { ...nonZeroBalances, ...changedBalances }
+
+    if (Object.keys(newBalancesState).length === 0) return
+    this.#pool.next(newBalancesState)
+  }
+
+  private setPool(balances: BalanceJson[]) {
+    const balancesWithIds = new Balances(balances).toJSON()
+    this.#pool.next(balancesWithIds)
+  }
+  /**
+   * Gets the balance for an address on a chain, either from the store if the address is in the wallet, or externally from the RPC if not.
+   *
+   * @param chainId - The id of the chain for which to query the balance
+   * @param evmNetworkId - The id of the evmNetwork for which to query the balance
+   * @param tokenId - The id of the token for which to query the balance
+   * @param address - The address to query the balance
+   */
+  async getBalance({
+    chainId,
+    evmNetworkId,
+    tokenId,
+    address: chainFormattedAddress,
+  }: RequestBalance): Promise<BalanceJson | undefined> {
+    assert(chainId || evmNetworkId, "chainId or evmNetworkId is required")
+
+    const address = encodeAnyAddress(chainFormattedAddress, 42)
+
+    // search for existing balance in the store
+    const storeBalances = new Balances(Object.values(this.balances))
+    const networkFilter = chainId ? { chainId } : { evmNetworkId }
```

Comment:
If someone calls `getBalance` with just a `tokenId` and `address`, the current `networkFilter` will limit the result to only substrate chains (balances where `evmNetworkId === undefined`)

I think removing the filter for this scenario would be less surprising:
```suggestion
    const networkFilter = chainId ? { chainId } : evmNetworkId ? { evmNetworkId } : {}
```

Reply:
Nevermind! Just noticed the `assert` at the top